ls1intum/Artemis
# Update OCaml template

### Checklist

### Motivation and Context
This year the FPV course wants to use Artemis with OCaml. The current OCaml template is rather simple and is missing some features that we would like to have.

### Description
- Prerequisite: Adding qcheck to the docker image in ls1intum/artemis-ocaml-docker#1
- Add an qcheck example test
- Add support for solution repository checkout
- Add support for not running some tests during the submission period
- Show build errors in student code instead of hiding all build output
- Disallow symlinks in the submission to prevent students from just including the solution
- Changing the template to make it easier for students to test and run their code with `dune utop src`


### Steps for Testing


1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to some test course
4. Create a new OCaml exercise
    Make sure to check the enable solution checkout box
5. ~~Open the test repository and check that all these files exist~~ *)
   <details><summary>find .</summary>
   <p>

   ```
   .
   ./.ocamlformat
   ./assignment
   ./assignment/assignment.mli
   ./assignment/dune
   ./.gitattributes
   ./.gitignore
   ./run.sh
   ./solution
   ./solution/dune
   ./solution/solution.mli
   ./test
   ./test/dune
   ./test/test.ml
   ```

   </p>
   </details>
6. Wait for the builds to complete and check that both template and solution build, that the template passes 0/8 tests and that the solution passes 7/8 tests
7. Change the variables HIDDEN_START and HIDDEN_END in run.sh in the tests repo to not include the current date
8. Wait for the solution build to finish and check that it now passes 8/8 tests
9. Cleanup the exercise and course, if required

### Review Progress





- Prerequisites
  - [ ] Review and merge ls1intum/artemis-ocaml-docker#1
  - [ ] Push the update docker image
- Code Review
- Manual Tests

### Test Coverage
I adjusted the excludes on the existing tests as far as i could tell what they were supposed to do and why ocaml/haskell were ex/included. There may be other places that i missed




### Notes
*) 
I hacked around to get artemis to create the repos without a CI, and confimed that it skips files without a dot and fixed the templates. The rest of this note is not really relevant for this PR anymore.

<hr/>

there are a bunch of special cases for template file names in https://github.com/ls1intum/Artemis/blob/6a7b15dc49cda4477cee17cfe622f2ec352f786a/src/main/java/de/tum/in/www1/artemis/service/FileService.java#L332-L354, but i am not sure why exactly and for what they are required.
While they appear to exist for all files that don't follow the `name.ext` format this doesn't appear to truly be necessary since the previous version of the template contained a `.merlin` file that was included in the template just fine without a special case.

The only potential issue i could spot is the pattern `*.*` in https://github.com/ls1intum/Artemis/blob/6a7b15dc49cda4477cee17cfe622f2ec352f786a/src/main/java/de/tum/in/www1/artemis/service/programming/ProgrammingExerciseService.java#L271-L273 which if the matcher doesn't special case to mean any file like on windows, could exclude files without a dot in the name. But that doesn't explain most of the special cases, or why the patterns isnt simply changed to `/**` or `/**/*`.

In the end i opted to leave files as is (other than gits files) and the just see what happens. I would have tests things locally but i don't have an bitbucket instance to test with and i don't know about artemis to make it work anyways.
